src/consts.rs

```diff
@@ -0,0 +1,396 @@
+        &ExprLit(ref lit) => Some(lit_to_constant(&lit.node)),
+        &ExprVec(ref vec) => constant_vec(cx, &vec[..]),
+        &ExprTup(ref tup) => constant_tup(cx, &tup[..]),
+        &ExprRepeat(ref value, ref number) =>
+            constant_binop_apply(cx, value, number,|v, n|
+                Some(ConstantRepeat(Box::new(v), n.as_u64() as usize))),
+        &ExprUnary(op, ref operand) => constant(cx, operand).and_then(
+            |o| match op {
+                UnNot =>
+                    if let ConstantBool(b) = o.constant {
+                        Some(Constant{
+                            needed_resolution: o.needed_resolution,
+                            constant: ConstantBool(!b),
+                        })
+                    } else { None },
+                UnNeg => constant_negate(o),
+                UnUniq | UnDeref => Some(o),
+            }),
+        &ExprBinary(op, ref left, ref right) =>
+            constant_binop(cx, op, left, right),
+        //TODO: add other expressions
+        _ => None,
+    }
+}
+
+fn lit_to_constant(lit: &Lit_) -> Constant {
+    match lit {
+        &LitStr(ref is, style) =>
+            Constant::new(ConstantStr(is.to_string(), style)),
+        &LitBinary(ref blob) => Constant::new(ConstantBinary(blob.clone())),
+        &LitByte(b) => Constant::new(ConstantByte(b)),
+        &LitChar(c) => Constant::new(ConstantChar(c)),
+        &LitInt(value, ty) => Constant::new(ConstantInt(value, ty)),
+        &LitFloat(ref is, ty) => {
+            Constant::new(ConstantFloat(is.to_string(), ty.into()))
+        },
+        &LitFloatUnsuffixed(ref is) => {
+            Constant::new(ConstantFloat(is.to_string(), FwAny))
+        },
+        &LitBool(b) => Constant::new(ConstantBool(b)),
+    }
+}
+
+/// create `Some(ConstantVec(..))` of all constants, unless there is any
+/// non-constant part
+fn constant_vec<E: Deref<Target=Expr> + Sized>(cx: &Context, vec: &[E]) -> Option<Constant> {
+    let mut parts = Vec::new();
+    let mut resolved = false;
+    for opt_part in vec.iter() {
+        match constant(cx, opt_part) {
+            Some(p) => {
+                resolved |= (&p).needed_resolution;
+                parts.push(p)
+            },
+            None => { return None; },
+        }
+    }
+    Some(Constant {
+        constant: ConstantVec(Box::new(parts)),
+        needed_resolution: resolved
+    })
+}
+
+fn constant_tup<E: Deref<Target=Expr> + Sized>(cx: &Context, tup: &[E]) -> Option<Constant> {
+    let mut parts = Vec::new();
+    let mut resolved = false;
+    for opt_part in tup.iter() {
+        match constant(cx, opt_part) {
+            Some(p) => {
+                resolved |= (&p).needed_resolution;
+                parts.push(p)
+            },
+            None => { return None; },
+        }
+    }
+    Some(Constant {
+        constant: ConstantTuple(Box::new(parts)),
+        needed_resolution: resolved
+    })
+}
+
+#[cfg(test)]
```

Comment:
Because I wanted to have two versions of the method, one for testing. However I have now decided that it's better to do some other trickery to keep the Context out of the picture.